text/0000-contract-metadata.md

```diff
@@ -0,0 +1,306 @@
+- Proposal Name: `contract_metadata`
+- Start Date: 2019-06-25
+- NEP PR: [nearprotocol/neps#0003](https://github.com/nearprotocol/NEPs/pull/3)
+
+# Summary
+[summary]: #summary
+
+This NEP introduces contract metadata, which summarizes the content of a given contract in json format.
+Contract metadata allows developers to easily list the methods in a contract and potentially display them to users to
+provide transparency. 
+It also provides the ability to validate input data in frontend thanks to the signatures of contract methods provided
+by the contract metadata.
+Contract metadata is similar to contract abi in some other blockchains such as Ethereum or Substrate but is more geared
+towards readability and developer friendliness.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there is no convenient way for a developer to programmatically list the methods of a contract that is deployed on chain
+because the contract code stored on chain is the compiled wasm code, where the parameters and type information are already mangled (even though
+method name is not, it is still cumbersome to extract them from wasm binary).
+As a result, if developers want to display the methods of a contract to an user of the app to provide transparency,
```

Comment:
As discussed with @bowenwang1996 before I think the only real application of the contract metadata is input validation. The case when front-end is somehow implemented to not have hardcoded dependencies on method or argument names has several issues:
* As discussed, it is very unlikely that front-end developer will somehow use a contract method without using the name of the method anywhere in their code (because they at least need to identify it) so if the contract method name changes they would have to manually modify the code anyway;
* It opens a huge security concern and such usage should be in general discouraged. We need to have a section in this NEP and documentation explaining the following example: Suppose someone malicious implements a contract with method `burn(gas_in_thousands: u64)` and lets other developers to use it in their front-ends. Then malicious contract developer changes the signature of the function to be `burn(gas: u64)` without announcing to anyone. If front-end did not break because it has automatically rebound to the the argument name the users now are burning 1000 more gas without having any notification. The core problem with metadata is that **it does not describe semantics, it describes syntax**

Reply:
> As discussed with @bowenwang1996 before I think the only real application of the contract metadata is input validation.

As well as it allows to generate a wrapper which would automatically encode/decode ABI (which could be considered as a validation too).